misc/package/Data Files/MWSE/core/lib/DependencyManager/v1/init.lua

```diff
@@ -0,0 +1,252 @@
+            message = string.format("%s\nRequired Version: %s", message, dependency.version)
+        end
+        if currentVersion then
+            message = string.format("%s\nCurrent Version: %s", message, currentVersion)
+        end
+        tes3ui.showMessageMenu{
+            header = "Dependency Failure",
+            message = message,
+            buttons = {
+                {
+                    text = "Okay",
+                    callback = function()
+                        tes3ui.leaveMenuMode()
+                    end,
+                    showRequirements = function()
+                        return dependency.url == nil
+                    end
+                },
+                {
+                    text = string.format("Download %s", dependency.name),
+                    callback = function()
+                        os.execute("start " .. dependency.url)
+                        os.exit()
+                    end,
+                    showRequirements = function()
+                        return dependency.url ~= nil
+                    end
+                },
+                {
+                    text = "Cancel",
+                    callback = function()
+                        tes3ui.leaveMenuMode()
+                    end,
+                    showRequirements = function()
+                        return dependency.url ~= nil
+                    end
+                }
+            }
+        }
+    end)
+end
+
+function DependencyManager:doLuaFileCheck(dependency)
+    local luaFile = include(dependency.luaFile)
+    if luaFile == nil then
+        self.logger:error("Could not find dependency file: %s", dependency.luaFile)
+        return false
+    else
+        if dependency.version then
+            local version = luaFile.version
+            if not version then
+                self.logger:error("Could not find version in dependency file: %s", dependency.luaFile)
+                return false
+            end
+            if not self:doVersionCheck(dependency, version) then
+                return false
+            end
+        end
+    end
+    return true
+end
+
+function DependencyManager:doVersionCheck(dependency, version)
+    if dependency.version == nil then
+        --no version required
+        return true
+    end
+    if version == nil then
+        -- version required but none found
+        return false
+    end
+    local major, minor, patch = string.match(version, "(%d+)%.(%d+)%.(%d+)")
+    self.logger:debug("Found version: Major: %s, Minor: %s, Patch: %s", major, minor, patch)
+    local targetMajor, targetMinor, targetPatch = string.match(dependency.version, "(%d+)%.(%d+)%.(%d+)")
+
+    local operator
+    --find one of possible operators at start of string
+    for operatorPattern, _ in pairs(self.operators) do
+        if string.startswith(dependency.version, operatorPattern) then
+            operator = operatorPattern
+            break
+        end
+    end
```

Comment:
I think this might not reliably detect > and >= at the beginning of the version string, since the table iteration order with `pairs()` in LuaJIT [is undefined](https://github.com/LuaJIT/LuaJIT/issues/702). 

So, there could be runs where it would iterate in this order: 
```
">="
">"
"="
```
And if the provided version were ">=1.0.0" you would end up with a ">" value in the operator variable.

One way you can work around that issue is to nest the operators in their table:
```Lua
	DependencyManager.operators = {
		{
			pattern = ">",
			callback = function() end,
		},		
		{
			pattern = ">=",
			callback = function() end,
		}
		{
			pattern = "=",
			callback = function() end,
		}
	},
```
Now that we have explicitly defined table order we can iterate over it using `ipairs()`:
```Lua
    local operatorCheck
    --find one of possible operators at start of string
    for operator, _ in ipairs(self.operators) do
        pattern = operator.pattern 
        if string.startswith(dependency.version, pattern ) then
            operatorCheck = operator.callback
            break
        end
    end
    -- and adjust the rest of the function
```

Reply:
It works fine as is, I have over 160 unit tests to prove it, and key lookups are faster than looping over the table (insignificant, I know, but why make it even marginally worse when it already works)

Reply:
https://github.com/MWSE/MWSE/blob/50d8f138f2f9687be8934bbb09e6721a830ea0c4/misc/package/Data%20Files/MWSE/core/lib/DependencyManager/v1/init.lua#L42